orchagent/routeorch.cpp

```diff
@@ -1227,14 +1227,46 @@ bool RouteOrch::addRoute(RouteBulkContext& ctx, const NextHopGroupKey &nextHops)
     sai_object_id_t& vrf_id = ctx.vrf_id;
     IpPrefix& ipPrefix = ctx.ip_prefix;
 
-    if (m_fgNhgOrch->fgNhgPrefixes.find(ipPrefix) != m_fgNhgOrch->fgNhgPrefixes.end()
-            && vrf_id == gVirtualRouterId)
+    if (m_fgNhgOrch->isRouteFineGrainedECMP(vrf_id, ipPrefix, nextHops))
     {
-        /* Only support the default vrf for Fine Grained ECMP */
-        SWSS_LOG_INFO("Reroute %s:%s to fgNhgOrch", ipPrefix.to_string().c_str(), 
-                nextHops.to_string().c_str());
+        /* The route needs Fine Grained ECMP, first check if routeorch has a route
+         * matching the prefix. If routeorch does then we need to delete the route
+         * so that it can be added for fine grained ecmp after that. */ 
+        auto route_table = m_syncdRoutes.find(vrf_id);
+        if (route_table != m_syncdRoutes.end())
+        {
+            auto route_entry = route_table->second.find(ipPrefix);
+            if (route_entry != route_table->second.end()) 
+            {
+                /* Case where we transition route from Regular ECMP to Fine Grained ECMP */
+                NextHopGroupKey nhgk = route_entry->second;
+                removeRoute(ctx);
+                gRouteBulker.flush();
+                if (!removeRoutePost(ctx))
+                {
+                    SWSS_LOG_ERROR("Failed to remove route %s from routeorch", ipPrefix.to_string().c_str());
+                    return false;
+                }
+                if (m_syncdNextHopGroups[nhgk].ref_count == 0)
+                {
+                    removeNextHopGroup(nhgk);
+                }
+            }
+        }
+        SWSS_LOG_INFO("Reroute %s:%s to fgNhgOrch", ipPrefix.to_string().c_str(),
+                        nextHops.to_string().c_str());
         return m_fgNhgOrch->addRoute(vrf_id, ipPrefix, nextHops);
     }
+    else
+    {
+        /* Route needs regular next-hop/ECMP. Logic here deals with case where we transition 
+         * route from Fine Grained ECMP to Regular ECMP: we first delete the route from fine grained ECMP 
+         * and then the rest of this function will create the route with regular ECMP/nexthops */
+        if (!m_fgNhgOrch->removeRoute(vrf_id, ipPrefix))
```

Comment:
Why to call this if it is not a FGNHG route? This is the else case, correct?

Reply:
This is for the case where there were Fine Grained next-hops a, b, c associated with prefix 1.1.1.1 and a route was created, now the route changes such that 1.1.1.1 points to regular ecmp next-hops d, e. When this happens, we need to remove the route from fgnhgorch and create it via routeorch in order to create regular ECMP.